brushfire-core/src/main/scala/com/stripe/brushfire/Tree.scala

```diff
@@ -2,38 +2,334 @@ package com.stripe.brushfire
 
 import com.twitter.algebird._
 
-// type Tree[K, V, T] = AnnotatedTree[K, V, T, Unit]
+sealed trait Node[K, V, T, A] {
+  def annotation: A
+}
 
-object Tree {
-  def apply[K, V, T](node: Node[K, V, T, Unit]): Tree[K, V, T] =
-    AnnotatedTree(node)
+case class SplitNode[K, V, T, A](
+    annotation: A,
+    children: Seq[(K, Predicate[V], Node[K, V, T, A])]) extends Node[K, V, T, A] {
+  require(children.nonEmpty)
+
+  def findChildren(row: Map[K, V]): List[Node[K, V, T, A]] =
+    children.collect { case (k, p, n) if p(row.get(k)) => n }(collection.breakOut)
+}
+
+object SplitNode{
+  def apply[K, V, T, A: Semigroup](children: Seq[(K, Predicate[V], Node[K, V, T, A])]): SplitNode[K, V, T, A] = {
+    require(children.nonEmpty)
+    val annotation = Semigroup.sumOption(children.view.map(_._3.annotation))
+    SplitNode(annotation.get, children)
+  }
+}
+
+case class LeafNode[K, V, T, A](
+  index: Int,
+  target: T,
+  annotation: A) extends Node[K, V, T, A]
+
+object LeafNode {
+  def apply[K, V, T, A: Monoid](index: Int, target: T): LeafNode[K, V, T, A] =
+    LeafNode(index, target, Monoid.zero)
+}
+
+case class Tree[K, V, T, A: Semigroup](root: Node[K, V, T, A]) {
+  private def mapSplits[K0, V0](f: (K, Predicate[V]) => (K0, Predicate[V0])): Tree[K0, V0, T, A] = {
+    def recur(node: Node[K, V, T, A]): Node[K0, V0, T, A] = node match {
+      case SplitNode(annotation, children) =>
+        SplitNode(annotation, children.map {
+          case (key, pred, child) =>
+            val (key0, pred0) = f(key, pred)
+            (key0, pred0, recur(child))
+        })
+
+      case LeafNode(index, target, annotation) =>
+        LeafNode(index, target, annotation)
+    }
+
+    Tree(recur(root))
+  }
+
+  private def mapLeaves[T0, A0: Semigroup](f: (T, A) => (T0, A0)): Tree[K, V, T0, A0] = {
```

Comment:
Might be a bit closer to a `fold` than a `map` now that `SplitNode` sums up the child annotations...